### PR TITLE
[MISC] Implemented 2021 Game State

### LINKED ISSUE #182: [MISC] Implement Spring Comp 2021 game states
Shepherd will send us a protobuf to signal what "game state" the robot is in. The behavior of Runtime during each game state is defined in the [game manual for Spring Comp 2021](https://docs.google.com/document/d/1248c_AExZhXhk0G51orj0MF8qQgVnIKR0KLEM4V1eNI/edit?usp=sharing).

Spring Comp 2021 Robot Details:
- Each robot will have three motors: Two for the wheels that move the robot and one for the end effector (a motor attached to an arm to interact with the environment.
 
Paraphrased from Shepherd:
- Poison Ivy: All motors will be reversed for ten seconds
   - Idea: The "invert" param on each motor of all KoalaBears will be the opposite of what it set by the student.
- Dehydration: All motors are stopped for 10 seconds.
  - Idea: Set the max velocity on all KoalaBears to 0.
- Hypothermia: All motors are slowed down.
  - Idea: Decrease the max velocity on all KoalaBears.

This may require changes in a variety of places, including shared memory, executor, lowcar, and/or dev handler.

The protobuf will be in this format:
```
/*
 * define an action runtime must take based on game state
 */
​
 syntax = "proto3";
​
 option optimize_for = LITE_RUNTIME;
 
 enum State {
     POISON_IVY = 0;
     DEHYDRATION = 1;
     HYPOTHERMIA_START = 2;
     HYPOTHERMIA_END = 3;
 }
 
 message GameState {
     State state = 1;
 }
 ```

### PR DESCRIPTION
**Pull request by benliao1**
Things that were specific to the 2021 game (`HYPOTHERMIA`, `POISON_IVY`, and `DEHYDRATION`) were implemented according to specifics from Shepherd and Atlas. There is a new file, `executor/filter.c`, which filters the commands to motors from student code according to the specification to create certain motor behaviors.

Changes were also made in `net_handler` and `shm_wrapper` to accomodate the new protobuf coming from Shepherd and to add additional robot state variables to keep track of these states.

Closes #182

### CODE REVIEW

File: tests/client/net_handler_client.c [hunk 1]
```diff
@@ -375,6 +376,41 @@ void send_run_mode(robot_desc_field_t client, robot_desc_val_t mode) {
...
+            }
+        default:
+            printf("ERROR: sending game state message\n");
+    }
+    len = game_state__get_packed_size(&game_state);
+    send_buf = make_buf(GAME_STATE_MSG, len);
+    game_state__pack(&game_state, send_buf + BUFFER_OFFSET);
+
+    // send the message
+    writen(nh_tcp_shep_fd, send_buf, len + BUFFER_OFFSET);
```
Comment (AVSurfer123) [inline tests/client/net_handler_client.c:409]: Need to check return status

File: executor/filter.c [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// Microseconds between each time we poll shared memory for the current game state
+#define GAMESTATE_POLL_INTERVAL (0.5 * 1000000)
+// Duration of POISON_IVY and DEHYDRATION in milliseconds
+#define DEBUFF_DURATION 10000
+// How much to slow motor velocities when HYPOTHERMIA is ACTIVE
+#define SLOW_SCALAR 0.2
+
+// Some "hardcoded" values to decrease overhead
+// device_name_to_type("KoalaBear")
+#define KOALABEAR 6
```
Comment (AVSurfer123) [inline executor/filter.c:18]: Shouldn't do this cus hardcoding is bad, instead just call the function in the beginning of `gamestate_handler`

File: executor/filter.c [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+// Parameter indices of the KoalaBear
+#define IDX_VELOCITY_A 0
+#define IDX_VELOCITY_B 8
```
Comment (AVSurfer123) [inline executor/filter.c:21]: Same for these. Instead use `get_param_idx`

File: executor/filter.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
+#include "filter.h"
```
Comment (AVSurfer123) [inline executor/filter.c:1]: Can we rename both files to `gamestate_filter.*` because filter is not descriptive

File: executor/Makefile [hunk 5]
```diff
@@ -10,7 +10,7 @@ LIBS=-pthread -lrt -export-dynamic -fPIC -lprotobuf-c
 
 all: executor studentapi.c
 
-executor: executor.c ../shm_wrapper/shm_wrapper.c ../logger/logger.c ../runtime_util/runtime_util.c ../net_handler/net_util.c ../net_handler/pbc_gen/text.pb-c.c
+executor: executor.c filter.c ../shm_wrapper/shm_wrapper.c ../logger/logger.c ../runtime_util/runtime_util.c ../net_handler/net_util.c ../net_handler/pbc_gen/text.pb-c.c
```
Comment (AVSurfer123) [inline executor/Makefile:13]: Rename this too

**Review by AVSurfer123 — CHANGES_REQUESTED** [review]

**Review by levincent06 — APPROVED** [review]

**Review by AVSurfer123 — APPROVED** [review]

**Pull request merged**